### PR TITLE
Fix buffer stride logic in ImageInput read methods

### PR DESCRIPTION
**Pull request by lgritz**
Reported in #2106, there was a subtle bug in read_image/scanline/tile
where we fell into the "contiguous and no data type translation"
case when that was not really the case.

The symptom could be triggered by the following case: You want to read a
1-channel float image into a buffer meant to hold a 4-channel UINT8
image with a 4-byte xstride. The correct behavior is to translate the
floats into uint8's and place them in channel 0 of each 4-channel pixel.
But because the buffer stride length (4 x uint8 = 4 bytes) matched the
native stride length (1 x float = 4 bytes), it was copying the floats
instead of translating and depositing them correctly in channel [0]
(leaving channels [1-3] unchanged).

The fix isn't too hard, but we made the same mistake in several places.
Basically, we were comparing whether the strides matched the native
spacing, instead of whether they matched the buffer spacing. Only
particular unusual combinations would trigger the false positive.

Add an explicit test of this case in imagebuf_test.cpp.

### CODE REVIEW

**Comment by lgritz** [comment]
Fixes #2106
